designs/2022-forbid-disable/README.md

```diff
@@ -22,48 +22,47 @@ Is a much worse rule to disable than
+The error on the inline config will have a severity of 2. See [LintMessage#severity](https://eslint.org/docs/developer-guide/nodejs-api#-lintmessage-type)
 
-note that when disabling all rules such as `/* eslint-disable */` the directive object will have `ruleId: null`. This will need a workaround. 
+first in `getDirectiveComments` I will match any directives that affect a rule listed in the array `noInlineConfig`. If there is a match, I will create a problem on that line to alert the user that they are attempting to disable a rule that will not be affected by the inline config.
 
 This error will **not** have a quick fix available
 
 
+In `applyDirectives` (`apply-disable-directives.js`) I will reference the problem. If that problem matches a name in the array `noInlineConfig` I will not apply any directives. This will ensure that `problems.push(problem)` gets run. Note that this same logic applies to configuration comments as well.
+
 ## Documentation
 
-We will want to update the documentation to reflect the new changes. The announcement can be integrated into the eslintrc update announcement.
+We will want to update the documentation ([here](https://eslint.org/docs/user-guide/configuring/rules#disabling-inline-comments)) to reflect the new changes. The announcement can be integrated into the config update announcement.
 
 ## Drawbacks
 
-this adds complexity to the already complex eslintrc experience. Users could be confused by the interaction with `forbid` and `noInlineConfig`.
+this adds complexity to the already complex config experience.
 
 This will have a very minor impact on perf.
 
+This design does not alert users in the case where they user `// eslint-disable` because that directive has `ruleId: null`. However, the design will still appropriately enforce the proper rules as desired.
+
 ## Backwards Compatibility Analysis
 
-This should integrate into the planned eslintrc changes and therefore will not be backwards compatible. However, this is acceptable since that will be a breaking change that is already approved and planned.
+This should integrate into the planned config changes and therefore will not be backwards compatible. However, this is acceptable since that will be a breaking change that is already approved and planned.
 
 ## Alternatives
 
-instead of adding a state to `error/warn` we could instead add another optional field to the rule config. This would enable a rule to be `warn` and have directives disabled. However, I believe that this actually adds more bloat to eslintrc which is why I chose the above design
+instead of extending `noInlineConfig` we could instead add a state called `enforce` to `error/warn`. This would be a superset of `error` where the rules cannot be disabled by inline config. However, it is hard to convey to the user succinctly what this accomplishes. And it allows users to disable the rules more easily in config files
```

Comment:
I do think there’s a user experience element that we need to consider between this and the noInlineConfig approach. 

Arguably, the “enforce” approach is more user-friendly because the request to disallow inline disabling is close to the rule configuration. I can see that as a plus.

The noInlineConfig option feels a little disassociated from the rule config and requires writing the rule name twice.

Reply:
There was some chatter on the original issues post. I think that keeping this away from the premier experience that we give `warn / error` is a good thing. It will keep users from abusing this feature. I want it to be an edge case for scenarios where you really need to protect a rule. Not front and center in the documentation. And it's so similar to `noInlineConfig` that it really feels like it fits

Reply:
I don’t consider that an edge case - every single application i use will likely restrict about 80% of the available rules once this feature is available.

Reply:
If repository owners end up trying to restrict most or all of the rules they use, then we're back to where we started, but arguably in a worse state because users can no longer opt out when they encounter legitimate edge cases.

Reply:
That would be the entire point - that to opt out, they’d have to make a change in the global code-owned eslint config.

Reply:
@nzakas  @bmish  @ljharb  I don't think that I will make more changes to the RFC until we come to a consensus on this. For what it's worth, changing ESLint config is easily restricted to required reviewers (the lint experts hopefully) in most large repos. So I see this as a strong protection against future changes. Also I disagree with @ljharb that this will be used for 80% of rules. @bmish rightfully pointed out that it only takes one edge case for each rule to be removed from the list. In my repo's I would only turn this feature on for rules such as [`no-restricted-disable`](https://mysticatea.github.io/eslint-plugin-eslint-comments/rules/require-description.html) So that people can't abuse the ESLint systems by disabling rules related to disabling.

So I stand by the current design as it reflects an out of the way option to have fine grain control of the Linter, without being a front and center use case

Reply:
In my experience, edge cases would be preferred to block progress rather than to warrant a disable line.

In other words, we have more faith in eslint rules than in human judgement; that's the point of automation.

Reply:
Maybe we could introduce a new, alternative way to configure individual rules:

```js
{
    rules: {
        semi: {
            severity: "error",
            options: ["never", { "beforeStatementContinuationChars": "always"}],
            noInlineConfig: true
        }
    }
}
```

Reply:
That would allow shared configs to define rules this way; is that desired?

Reply:
Allowing shared configs to include `noInlineConfig` (i.e. controlling which rules users can easily inline-disable) sounds pretty extreme and undesirable to me. Shared config owners naturally consider their own rules to be of heightened importance and so I'd worry we would end up with a bunch of shared configs over-aggressively enabling `noInlineConfig`. Someone else choosing which rules I can inline-disable in my repositories seems like it would take control away from me about how I use ESLint and also lead to confusion with users wondering why they can disable certain rules inline but not others.

Reply:
There is no difference between shareable configs and local configs in ESLint. Anything you can configure in one may also be configured in the other. I don’t see us introducing exceptions to this approach anytime soon.

I do strongly prefer keeping configuration of a rule close to that rule in the config. Whether that be a new severity like “enforce”, a new combo severity like “error+enforce”, or the approach @mdjermanovic proposed with an expanded rule config object, I do think that’s preferable to needing to enter the rule ID in a different location.

To answer some other concerns:

1. I don’t think it’s important who is allowed to do what. Whoever has control over a project’s config should be allowed to do whatever they want. We aren’t in the business of preventing maintainers from implementing ridiculously strict configs. Teams can fight that out on their own. 
2. We aren’t in a position to know whether disabling inline config on a per-rule level is an edge case or not right now. I don’t think we should come at this from a perspective of how common it is. Instead, let’s assume it will be common and figure out what is the least painful way to do it.

Reply:
Okay I updated it to go back to the `enforce` design. I do feel as if there are a lot of voices coming in here so it makes it hard to parse what the consensus of the group is on this feature and the design options. I'm happy to attend a TSC meeting to discuss as well if that would be beneficial. @nzakas @ljharb  @bmish  @mdjermanovic

Reply:
RFCs are open to public comment, so you may not get consensus. Generally what we are going for is a consensus among the TSC. You can see who’s on the team at eslint.org/team.

Reply:
Okay got it. I will admit that I'm having a hard time grasping what the next steps for the RFC will be. I can keep changing the design doc, but I worry that the issue isn't the design but rather peoples desire to have this be added at all. So for that reason I will leave it as is until I get a few more comments and direction from folks. Then we can discuss if this is a candidate for acceptance.

Reply:
Yup, that’s exactly why those sections exist. Just in case we need to revisit the design in the future, having those documented is really helpful.